utils/utils.go

```diff
@@ -14,101 +14,238 @@ limitations under the License.
 package utils
 
 import (
+	"archive/zip"
+	"io"
 	"io/ioutil"
+	"log"
 	"net/http"
+	"os"
+	"path/filepath"
 
 	pkgerrors "github.com/pkg/errors"
 	appsV1 "k8s.io/api/apps/v1"
 	coreV1 "k8s.io/api/core/v1"
 	"k8s.io/client-go/kubernetes/scheme"
 )
 
-// Download the raw CSAR file from a specific URL
-var Download = func(url string) ([]byte, error) {
+var CSAR FileOperator
+
+// FileOperator is an interface to Download, Extract and Delete files
+type FileOperator interface {
+	Download(string, string) error
+	Unzip(string) error
+	Delete(string) error
+}
+
+// CSARFile is a concrete type to Download, Extract and Delete CSAR files
+type CSARFile struct{}
+
+// Download a CSAR file
+func (c *CSARFile) Download(fileName string, url string) error {
+	file, err := os.Create(fileName)
+	if err != nil {
+		return pkgerrors.Wrap(err, "CSAR file create error")
+	}
+	defer file.Close()
+
 	resp, err := http.Get(url)
 	if err != nil {
-		return nil, pkgerrors.Wrap(err, "Get CSAR file error")
+		return pkgerrors.Wrap(err, "CSAR file download error")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return pkgerrors.Wrap(err, "CSAR file download error")
 	}
 	defer resp.Body.Close()
 
-	rawBody, err := ioutil.ReadAll(resp.Body)
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return pkgerrors.Wrap(err, "CSAR file encode error")
+	}
+
+	return nil
+}
+
+// Unzip a CSAR file
+func (c *CSARFile) Unzip(filePath string) error {
```

Comment:
This method can be declared as private

Reply:
Since it has to satisfy the interface https://github.com/shank7485/k8-plugin-multicloud/pull/13/files#diff-cd575165f84eed10d10eb1460b0927b4R36, I have to make them public.